src/legacy/core_plugins/vis_default_editor/public/components/controls/raw_json.tsx

```diff
@@ -34,48 +32,77 @@ function RawJsonParamEditor({
   setValue,
   setTouched,
 }: AggParamEditorProps<string>) {
-  const label = (
-    <>
-      <FormattedMessage id="visDefaultEditor.controls.jsonInputLabel" defaultMessage="JSON input" />{' '}
-      <EuiIconTip
-        position="right"
-        content={i18n.translate('visDefaultEditor.controls.jsonInputTooltip', {
-          defaultMessage:
-            "Any JSON formatted properties you add here will be merged with the elasticsearch aggregation definition for this section. For example 'shard_size' on a terms aggregation.",
-        })}
-        type="questionInCircle"
-      />
-    </>
+  const [isFieldValid, setFieldValidity] = useState(true);
+  const [editorReady, setEditorReady] = useState(false);
+
+  const editorTooltipText = useMemo(
+    () =>
+      i18n.translate('visDefaultEditor.controls.jsonInputTooltip', {
+        defaultMessage:
+          "Any JSON formatted properties you add here will be merged with the elasticsearch aggregation definition for this section. For example 'shard_size' on a terms aggregation.",
+      }),
+    []
   );
-  const isValid = isValidJson(value);
 
-  const onChange = (ev: React.ChangeEvent<HTMLTextAreaElement>) => {
-    const textValue = ev.target.value;
-    setValue(textValue);
-    setValidity(isValidJson(textValue));
-  };
+  const jsonEditorLabelText = useMemo(
+    () =>
+      i18n.translate('visDefaultEditor.controls.jsonInputLabel', {
+        defaultMessage: 'JSON input',
+      }),
+    []
+  );
 
-  useEffect(() => {
-    setValidity(isValid);
-  }, [isValid]);
+  const label = useMemo(
+    () => (
+      <>
+        {jsonEditorLabelText}{' '}
+        <EuiIconTip position="right" content={editorTooltipText} type="questionInCircle" />
+      </>
+    ),
+    []
+  );
+
+  const onEditorValidate = useCallback(
+    (annotations: unknown[]) => {
+      // The first onValidate returned from EuiCodeEditor is a false negative
+      if (editorReady) {
+        const validity = annotations.length === 0;
+        setFieldValidity(validity);
+        setValidity(validity);
+      } else {
+        setEditorReady(true);
+      }
+    },
+    [setValidity]
+  );
 
   return (
     <EuiFormRow
       label={label}
-      isInvalid={showValidation ? !isValid : false}
+      isInvalid={showValidation ? !isFieldValid : false}
       fullWidth={true}
       compressed
     >
-      <EuiTextArea
-        id={`visEditorRawJson${agg.id}`}
-        isInvalid={showValidation ? !isValid : false}
-        value={value}
-        onChange={onChange}
-        rows={2}
-        fullWidth={true}
-        onBlur={setTouched}
-        compressed
-      />
+      <>
+        <EuiCodeEditor
+          mode="json"
+          theme="github"
+          width="100%"
+          height="250px"
+          value={value}
+          onValidate={onEditorValidate}
+          setOptions={{
+            fontSize: '14px',
```

Comment:
This is the only line, concerning me a bit :-)
We set this manually to make the font larger or smaller?

@cchaos do you know if we can somehow import from EUI here?

Reply:
This is actually exactly how we're doing it in the docs, so really I think it's fine. You **could** pull in the value from our JSON file, if that's the worry. Really this component **should** get sunsetted in favor of the Monaco editor.